client/asset/dcr/dcr.go

```diff
@@ -2846,6 +2942,60 @@ func (dcr *ExchangeWallet) getBestBlock(ctx context.Context) (*block, error) {
 	return &block{hash: hash, height: height}, nil
 }
 
+func (dcr *ExchangeWallet) getBlock(ctx context.Context, blockHash *chainhash.Hash, verboseTx bool) (*chainjson.GetBlockVerboseResult, error) {
+	blockVerbose, err := dcr.wallet.GetBlockVerbose(ctx, blockHash, verboseTx)
+	if err != nil {
+		return nil, fmt.Errorf("error retrieving block %s: %w", blockHash, translateRPCCancelErr(err))
+	}
+	return blockVerbose, nil
+}
+
+func (dcr *ExchangeWallet) getBlockHash(blockHeight int64) (*chainhash.Hash, error) {
+	blockHash, err := dcr.wallet.GetBlockHash(dcr.ctx, blockHeight)
+	if err != nil {
+		return nil, translateRPCCancelErr(err)
+	}
+	return blockHash, nil
+}
+
+// mainchainAncestor crawls blocks backwards starting at the provided hash
+// until finding a mainchain block. Returns the first mainchain block found.
+func (dcr *ExchangeWallet) mainchainAncestor(ctx context.Context, blockHash *chainhash.Hash) (*chainhash.Hash, int64, error) {
+	checkHash := blockHash
+	for {
+		checkBlock, err := dcr.wallet.GetBlockHeaderVerbose(ctx, checkHash)
+		if err != nil {
+			return nil, 0, fmt.Errorf("getblockheader error for block %s: %w", checkHash, translateRPCCancelErr(err))
+		}
+		if checkBlock.Confirmations > -1 {
+			// This is a mainchain block, return the hash and height.
+			return checkHash, int64(checkBlock.Height), nil
+		}
+		if checkBlock.Height == 0 {
+			return nil, 0, fmt.Errorf("no mainchain ancestor for block %s", blockHash.String())
+		}
+		checkHash, err = chainhash.NewHashFromStr(checkBlock.PreviousHash)
+		if err != nil {
+			return nil, 0, fmt.Errorf("error decoding previous hash %s for block %s: %w",
+				checkBlock.PreviousHash, checkHash.String(), err)
+		}
+	}
+}
+
+func (dcr *ExchangeWallet) isMainchainBlock(ctx context.Context, block *block) (bool, error) {
+	blockHeader, err := dcr.wallet.GetBlockHeaderVerbose(ctx, block.hash)
+	if err != nil {
+		return false, fmt.Errorf("getblockheader error for block %s: %w", block.hash, translateRPCCancelErr(err))
+	}
+	return blockHeader.Confirmations > -1 && int64(blockHeader.Height) == block.height, nil
```

Comment:
Just noting that now that we don't just ask dcrd via dcrwallet for everything, we're also taking on the burden of accounting for stake invalidation too.  "invalidated" in callers so far is just being moved to a side chain.  Lets move ahead like you have, but we would really need to look at the next block's votebits to see if this one has had its regular tx tree invalidated.  :(

Reply:
Hmm. Should be a quick fix. Will add.